src/glimt/multi_state.cljc

```diff
@@ -0,0 +1,211 @@
+                                                                   :on    {::error   [{:guard  more-retries?
+                                                                                       :target ::waiting}
+                                                                                      {:target  (or failure-state (vec (concat state-path [::error ::halted])))
+                                                                                       :actions (dispatch-callback :on-failure)}]
+                                                                           ::success [{:target  (or success-state (vec (concat state-path [::loaded])))
+                                                                                       :actions dispatch-success}]}}}}
+                                  ::halted   {}}}
+             ::loaded  {}}})
+
+;; Register a `machine`, saving it at `fsm-path`. This is merely a convenience
+;; alias for `::sc.rf/register`. It can be useful to avoid an extra require if
+;; you need to register a machine that has a glimt embedded machine within it.
+;; After a machine has been registered, start a request for it by dispatching
+;; `::http/start` with the `fsm-path` provided here as the request's
+;; `:fsm-path`.
+(f/reg-event-fx
+ ::register
+ (fn [_ [_ fsm-path machine]]
+   {:dispatch [::sc.rf/register fsm-path machine]}))
+
+;; Register the default FSM. If a request doesn't include an `:fsm-path`, this
+;; FSM will be used.
+(let [machine (assoc (embedded-fsm {:state-path [:>]}) :id default-fsm-id)]
+  (f/dispatch [::register default-fsm-path (sc/machine machine)]))
+
+(defn request-state-path [request-id]
+  (into [::requests] (sc.utils/ensure-vector request-id)))
+
+(defn- request-transition-data [{:keys [id fsm-path]}]
+  {:fsm-path   (or fsm-path default-fsm-path)
+   :state-path (request-state-path id)})
+
+;; Helper for requests that specify a `:path` instead of an `:on-success`.
+(f/reg-event-db ::save-to-path (fn [db [_ path data]] (assoc-in db path data)))
+
+;; Start a request, storing it and its state in the DB.
+(f/reg-event-fx
+ ::start
+ (fn [_ [_ request]]
+   (if-let [errors (m/explain request-validation-schema request)]
+     {::sc.rf/log [:error "Invalid HTTP request" (str (me/humanize errors)) errors]}
+     (let [transition-data (request-transition-data request)
+           request         (-> (m/decode request-schema request mt/default-value-transformer)
+                               ;; retry-delay is a function (of how many retries have been
+                               ;; performed) or an int. Coerce it to be a function.
+                               (update :retry-delay
+                                       (fn [retry-delay]
+                                         (if (int? retry-delay)
+                                           (constantly retry-delay)
+                                           retry-delay)))
```

Comment:
Quick random idea (not super important):

Is it possible and practical to have this coercion as part of malli decoding?

Reply:
Possible, yes. But the only way I came up with co-opts the `m/string-transformer` into doing something it wasn't designed to do:

```clojure
(def request-schema-def

...

   [:retry-delay {:default       2000
                  :decode/string (fn [retry-delay]
                                   (if (ifn? retry-delay)
                                     retry-delay
                                     (constantly retry-delay)))}

...

(m/decode request-schema request (mt/transformer mt/default-value-transformer mt/string-transformer))
```

I think the other option is to write a custom transformer, though I don't really know how to do that.

I'm divided on using malli for decoding. I know that's one of the features that distinguishes it from clojure.spec.alpha. But I find transformations written in regular clojure much easier to reason about. In order to understand the supposedly "declarative" transformations in the schema, I have to know something about how transformers work (including interceptor chains, schema properties, etc.)

Reply:
Well, both the possible and practical part lands on a "no" here, I guess :) No big deal, just was curious.

Decoding is nice for filling in default values, and also if there is a clear and natural fit. In this case it isn't.